napari/layers/utils/layer_utils.py

```diff
@@ -204,81 +222,470 @@ def _nanmax(array):
-            ]
-        # compute everything in one go
-        reduced_data = dask.compute(*reduced_data)
+        # Indices are either numpy array or chunk indices dependent on data structure of data and are 0 < length <= 3
+        idxs = _get_plane_indices(shape, offset)
+        slices_2D = _get_crop_slices(shape, idxs, offset, chunk_size)
+
+        reduced_data = [
+            [_nanmax(data[sl]) for sl in slices_2D],
+            [_nanmin(data[sl]) for sl in slices_2D],
+        ]
     else:
         reduced_data = data
 
+    if chunk_size:
+        reduced_data = dask.compute(*reduced_data)
+
     min_val = _nanmin(reduced_data)
     max_val = _nanmax(reduced_data)
 
     if min_val == max_val:
         min_val = 0
         max_val = 1
-    return (float(min_val), float(max_val))
+    return float(min_val), float(max_val)
+
+
+def _get_1d_slices(shape, chunk_size) -> list[slice]:
+    """
+    Calculate data range in case of 1D data.
+
+    Parameters
+    ----------
+    shape: Sequence[int]
+        Shape of the data either in pixels or chunks.
+    chunk_size: Sequence[int]
+        The size in pixels per dimension of the chunks if data is lazy.
+
+    Returns
+    -------
+    slices: list[slice]
+        Slices to be taken from data for calculating contrast limits.
+    """
+    # If data is very large we only aim to take data around the center.
+    center = shape[0] // 2 * chunk_size[0] if chunk_size else shape[0] // 2
+
+    if not chunk_size:
+        return [
+            slice(
+                center - int(PIXEL_THRESHOLD // 2),
+                center + int(PIXEL_THRESHOLD // 2),
+            )
+        ]
+
+    chunk_size_product = np.prod(chunk_size)
+    allowed_chunks = int(PIXEL_THRESHOLD // chunk_size_product)
+    multiplier = min(allowed_chunks, MAX_NUMBER_OF_CHUNKS) // 2
+
+    return [
+        slice(
+            center - chunk_size[0] * multiplier,
+            center + chunk_size[0] * multiplier,
+        )
+    ]
+
+
+def _get_blocks_grid_shape(
+    data_shape: Sequence[int], chunk_size: Sequence[int]
+) -> Tuple[int, ...]:
+    """
+    Get the block shape: the approximate layout of the grid of array chunks.
+
+    Gets the shape of the array not on the pixel level but on the chunk level. For example if x is 100 and
+    chunk size is 10, 10 chunks fit in 100 and will thus be returned. In case of x being 105, still 10 will be
+    returned. This to ensure that only whole chunks will be loaded.
+
+    Parameters
+    ----------
+    data_shape: tuple[int, ...]
+        The shape of an array of chunked data.
+    chunk_size: tuple[int, ...]
+        The size per dimension of the chunks in the chunked data array.
+
+    Returns
+    -------
+    tuple[int, ...]
+        shape indicating the number of chunks per dimension.
+    """
+    return tuple(shape // size for shape, size in zip(data_shape, chunk_size))
+
+
+def _get_plane_indices(
+    shape: Sequence[int], offset: int
+) -> List[Tuple[int, ...]]:
```

Comment:
```suggestion
) -> list[tuple[int, ...]]:
```

typing annotation should allow for this